src/react/useQueries.ts

```diff
@@ -0,0 +1,137 @@
+import * as React from 'react';
+import { Monoid } from 'fp-ts/lib/Monoid';
+import { Option, fromNullable } from 'fp-ts/lib/Option';
+import { QueryResult } from '../QueryResult';
+import { product, ObservableQuery } from '../Query';
+import {
+  EnforceNonEmptyRecord,
+  ObservableQueries,
+  ProductL,
+  ProductP,
+  ProductA,
+  VoidInputObservableQueries
+} from '../util';
+import { defaultMonoidResult } from './util';
+import { observable } from '../Observable';
+import { observeShallow } from '../observe';
+import { setoidShallow } from '../Strategy';
+
+function usePrevious<T>(value: T): Option<T> {
+  const ref = React.useRef<T>();
+  React.useEffect(() => {
+    ref.current = value;
+  });
+  return fromNullable(ref.current);
+}
+
+/**
+ * A React hook to observe an `ObservableQuery`
+ *
+ * @param query an `ObservableQueries`
+ * @param params input values for the query
+ * @param resultMonoid an optional monoid used to aggregate `QueryResult`s
+ * @returns the latest `QueryResult` to operate upon
+ *
+ * @example
+ * useQuery(randomQuery, randomQueryParams).fold(
+ *   () => "load",
+ *   () => "failed",
+ *   (randomQueryResult) => randomQueryResult
+ * )
+ */
+export function useQuery<A extends void, L, P>(
+  query: ObservableQuery<A, L, P>,
+  params?: A,
+  resultMonoid?: Monoid<QueryResult<L, P>>
+): QueryResult<L, P>;
+export function useQuery<A, L, P>(
+  query: ObservableQuery<A, L, P>,
+  params: A,
+  resultMonoid?: Monoid<QueryResult<L, P>>
+): QueryResult<L, P>;
+export function useQuery<A, L, P>(
+  query: ObservableQuery<A, L, P>,
+  params: A,
+  resultMonoid?: Monoid<QueryResult<L, P>>
+): QueryResult<L, P> {
```

Comment:
aren't these two overloads identical?

Reply:
the "overload list" as seen by TS and users is only the list of signatures without implementation. The last signature (with implementation) can be as loose as desired, as long as it is compatible with any other signature in the list, but it is not "exported". I usually try to be as strict as possible in order to get help in the implementation. See e.g. https://www.typescriptlang.org/docs/handbook/functions.html#overloadshttps://www.typescriptlang.org/docs/handbook/functions.html#overloads

Reply:
ah, TIL!